### PR TITLE
parsl.executors.FluxExecutor

### PR DESCRIPTION
**Pull request by jameshcorbett**
# Description

#2050 

Adds an executor which leverages the Flux resource manager to launch apps and return futures.

Fixes # (2050)

## Type of change

- New feature (non-breaking change that adds functionality)

### CODE REVIEW

File: parsl/app/bash.py
Comment (jameshcorbett) [reply parsl/app/bash.py:68]: Yeah, it will affect all executors in theory. In practice, though, do they check all their open fds, or do something else that would be affected by this change? 

The problem is that Flux provides MPI to jobs by letting them inherit a FD and setting an environment variable defining what the FD is. So in `MPI_Init` the MPI implementation will read the environment variable and go to open the FD, which fails if `close_fds=True`. 

There may be other ways to provide MPI, though, so the problem is specific to that specific method of providing it, and to executors that launch MPI jobs.

Comment (benclifford) [reply parsl/app/bash.py:68]: I think its fine - was interested in having the reasoning recorded here in case something interesting shows up later.

**Comment by benclifford** [comment]
There are a few problems with mypy type checking. From easiest to hardest:

Because flux isn't installed in the test environment, tell mypy to stop trying to typecheck flux modules - eg in mypy.ini, add:

```
[mypy-flux.*]
ignore_missing_imports = True
```

(or if you are intending to make CI test the flux executor, then this would go away when you install that)

HighThroughputExecutor's process worker pool doesn't pass mypy type checking - prior to this PR, it was ignored by mypy so didn't cause failures, but this PR imports code from inside process_worker_pool to re-use. That makes me a little uncomfortable.

In another branch, I have a bunch of typechecking work which fixes the process worker pool to typecheck properly - your PR is motivation for me to get that piece put on master, so I will make a separate PR for that now.

**Comment by benclifford** [comment]
> In another branch, I have a bunch of typechecking work which fixes the process worker pool to typecheck properly - your PR is motivation for me to get that piece put on master, so I will make a separate PR for that now.

That PR, #2052, is merged into `master` now

**Comment by jameshcorbett** [comment]
Thanks---rebased and pushed.

**Comment by jameshcorbett** [comment]
I told mypy to stop checking Flux modules like you suggested. I do think it would be good to make CI test the executor. I think all we would need to do is install Flux and then get the tests running inside of a Flux instance. I could write some tests that assume an active Flux instance, or at least that Flux is in the PATH. Would it be feasible then to get Flux installed on the machine used for CI?

From a quick look just now, the CI seems to have failed just now for other reasons that don't have anything to do with my changes.

**Comment by jameshcorbett** [comment]
I don't know how much everyone is up on Flux, so I will try to explain the executor. Every app submitted to the executor is run as an individual Flux job. Although technically, each app is wrapped up in that `execute_parsl_task` script that I added in my first commit, and that script is what is actually submitted as the Flux job.

(For bash apps I think it would be more efficient to just forget the wrapper script and submit the bash app directly to Flux. Any thoughts on whether that is a good idea and feasible? It would break a lot of the abstraction.)

Flux jobs are fairly heavyweight. I think with the latest Flux version, I've seen no more than ~50 jobs/sec on our clusters, regardless of resources, and I think that's pretty normal. So at the moment, I don't think the FluxExecutor is a great fit for users who want to run tons and tons of fast-running jobs.

There is a big caveat to the performance numbers, however. It's possible to improve Flux's performance almost infinitely by duplicating it. This comes at two big costs: resource fragmentation and code complexity, because you (usually) need a way of communicating and load-balancing between all the different Flux instances. I think Parsl may already have most of this logic, though? In any case, the current executor I have here doesn't do any duplication---there's just a single Flux instance.

For a use-case that consists primarily of small, fast python functions, I don't think the FluxExecutor is a good fit because I suspect you already have better executors for that kind of workflow. I'm not sure what Flux would add to that use case. But if you want to maximize Flux's performance I will need to do a lot more work to implement the load-balancing and duplicating and so on.

Right now, however, where I think the FluxExecutor would be most useful would be on apps that are a little slower and need a dedicated and nontrivial resource set (like MPI jobs), and especially when the resource requirements vary across apps. Flux (and the FluxExecutor here) would handle that kind of app no problem.

The resource options for each app (accepted keys in the `parsl_resource_specification`) are:
- num_tasks: the number of tasks to launch (MPI ranks for an MPI job), default 1
- cores_per_task: cores per task, default 1
- gpus_per_task: gpus per task, default 1
- num_nodes: if > 0, evenly distribute the allocated cores/gpus across N nodes. Does _not_ give the job exclusive access to those nodes. This option only affects distribution.

As for the availability of Flux, it is fairly portable and easy to build, and it is currently used on LLNL and ORNL machines running Slurm and LSF. So for Parsl, this executor wouldn't necessarily just be an LLNL-Slurm-only executor or something like that. However, Flux has known problems on Cray machines (like ALCF's Theta or LANL's Trinitite/Trinity) because of vendor-specific extensions to Cray's MPI libraries. The consequence is that if you want to run an MPI job under Flux on those machines, as of right now the only way to do it is to compile the MPI application with OpenMPI rather than the default Cray MPI. I am not sure when that problem is likely to be resolved, if at all.

-------------
Sorry for the information dump, I just want to make sure that everyone on the Parsl team knows what they're getting and what they're _not_ getting. If people are happy with what I've implemented now and just want some tests and that's it, I could do that no problem.

Tagging @kylechard @yadudoc and @hategan in case they can help figure out the path forward.

**Comment by benclifford** [comment]
> (For bash apps I think it would be more efficient to just forget the wrapper script and submit the bash app directly to Flux. Any thoughts on whether that is a good idea and feasible? It would break a lot of the abstraction.)

There are various parsl features that rely on running python code on the worker: some file staging, monitoring, time limiting. 

There's also a theoretical possibility (but I've never seen it in real usage) that the python bit of a bash app looks at its environment to decide what bash command to run. That would also break - but I've never seen this done in real life.

The whole execution environment is built around executing python functions (some of which happen to fork out as bash apps) - that's pretty fundamental to everything parsl does as a python library for remotely executing python code. I think putting in deep changes to turn parsl core into something which can also run command lines, not python code, is a direction that we shouldn't be pursuing.

> Flux jobs are fairly heavyweight. I think with the latest Flux version, I've seen no more than ~50 jobs/sec on our clusters, regardless of resources, and I think that's pretty normal. So at the moment, I don't think the FluxExecutor is a great fit for users who want to run tons and tons of fast-running jobs.

I think thats fine -- things like the work queue executor are relatively heavy in their per-task startup, and make up for it by being better than parsl's earlier executors in other ways - primarily through resource allocation. I think its important to be clear about the costs and benefits and "why should you choose executor A over executor B" for all of our executors, and you've written text for that in this PR.

> There is a big caveat to the performance numbers, however. It's possible to improve Flux's performance almost infinitely by duplicating it. This comes at two big costs: resource fragmentation and code complexity, because you (usually) need a way of communicating and load-balancing between all the different Flux instances. I think Parsl may already have most of this logic, though? In any case, the current executor I have here doesn't do any duplication---there's just a single Flux instance.

There's nothing that does any load balancing between multiple executors in parsl data flow kernel: it sends tasks to an executor, and lets the executor do its own magic to get tasks executed.

**Comment by jameshcorbett** [comment]
I forgot to mention that each Future is associated with several files---another reason the executor shouldn't be used for large numbers of futures.

> The whole execution environment is built around executing python functions (some of which happen to fork out as bash apps) - that's pretty fundamental to everything parsl does as a python library for remotely executing python code. I think putting in deep changes to turn parsl core into something which can also run command lines, not python code, is a direction that we shouldn't be pursuing.

OK, good to know so that possibility can be ruled out.

> There's nothing that does any load balancing between multiple executors in parsl data flow kernel: it sends tasks to an executor, and lets the executor do its own magic to get tasks executed.

In that case, it sounds like as a future optimization, the executor could be turned into a front-end for multiple sub-executors, and the top-level executor (the only executor exposed to the Parsl data flow kernel) would just be responsible for load balancing between them. For certain workflows that would improve throughput enormously, but at the moment, given current Flux APIs, it would require that the user specify:

1. The number of partitions or sub-executors (technically, the full hierarchy of Flux instances, like '5x6x12', but for simplicity let's just say the number of partitions/sub-executors).
2. The number of nodes available to Flux
3. The number of cores and gpus per node

File: parsl/executors/flux/execute_parsl_task.py
Comment (jameshcorbett) [inline parsl/executors/flux/execute_parsl_task.py:31]: As it stands, there is no way to distinguish between an exception thrown by a callable and one returned. So if the callable raises an exception, you would fetch it with ``future.result()`` and not ``future.exception()`` as one might expect. Is this behavior consistent with the other Parsl executors? I'm guessing not, just want to make sure.

File: parsl/executors/flux/executor.py
Comment (jameshcorbett) [inline parsl/executors/flux/executor.py:184]: Should stdout and stderr be separated?

File: parsl/executors/flux/executor.py
Comment (jameshcorbett) [inline parsl/executors/flux/executor.py:179]: What should be the CWD for the apps?

File: parsl/executors/flux/executor.py
Comment (jameshcorbett) [inline parsl/executors/flux/executor.py:154]: Flux isn't the default resource manager on any systems yet, so currently the primary way to use Flux is to request an allocation from the local resource manager (like Slurm or LSF) and then launch Flux within that allocation (e.g. srun --ntasks-per-node=1 --mpi=none flux start COMMAND``). How would that work out with the various provides offered by Parsl? I forced ``LocalProvider`` to require that users already have acquired their resource allocation and launched Flux, but that cuts down dramatically on the potential use-cases for this executor.

File: parsl/executors/flux/executor.py
Comment (jameshcorbett) [inline parsl/executors/flux/executor.py:101]: Currently the class inherits from NoStatusHandlingExecutor for convenience only. Flux provides status updates on all submitted jobs, so it would be pretty trivial to provide that on this executor. However, I see that ``Executor.status()`` is supposed to return a mapping from strings to ``JobStatus`` objects. What are the strings supposed to be? I already have a uuid for each Future returned from the executor, so it would make sense to use that string. However, how is the user supposed to figure out which uuid goes with which Future?

File: parsl/executors/flux/executor.py
Comment (jameshcorbett) [reply parsl/executors/flux/executor.py:223]: One option would be to accept the CWD and environment through the resource spec.

File: parsl/executors/flux/executor.py
Comment (jameshcorbett) [inline parsl/executors/flux/executor.py:205]: I wonder if it might be easier to use and debug if the task IDs were 0, 1, 2, 3, ..., and every executor instance was given a unique workdir (so, unique workdirs instead of unique file names). As it stands, with all the files produced, it can be very hard to remember which app/future goes with which UUID, because the UUIDs are so long. There would just be less to memorize.

**Comment by jameshcorbett** [comment]
Dropping the draft tag. I left a bunch of inline comments where I need someone more familiar with Parsl to help out. Otherwise, and aside from the Flux hierarchy performance thing I mentioned above (which should be left for another time), as far as I know this is ready to go.

File: parsl/executors/flux/execute_parsl_task.py
Comment (yadudoc) [reply parsl/executors/flux/execute_parsl_task.py:31]: If the call raised, `future.result()` will raise the same exception. This is consistent with the `concurrent.futures` behavior: https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future.result. `future.exception()` on the other hand returns the exception and does not raise it.

File: parsl/executors/flux/executor.py
Comment (yadudoc) [reply parsl/executors/flux/executor.py:101]: Scaling and status handling in Parsl is fairly messy. The executor doesn't need to be a StatusHandlingExecutor if we don't expect to plug into Parsl's mechanism to gauge available resource (nodes/compute) and the current task load and then try to request through the executor to get additional nodes. Ultimately the question is whether we want any dynamic resource provisioning and if so whether Flux or Parsl should be making the provisioning decisions.

Comment (yadudoc) [reply parsl/executors/flux/executor.py:101]: If we just want a fixed-sized set of resources (the equivalent of 1 batch job) and just schedule tasks within it, `NoStatusHandlingExecutor` is probably the right choice.

File: parsl/executors/flux/execute_parsl_task.py
Comment (jameshcorbett) [reply parsl/executors/flux/execute_parsl_task.py:31]: Right, that's what I thought. I think I inherited this code from somewhere else and just wanted to make sure it worked correctly. Thanks for clarifying!

File: parsl/executors/flux/executor.py
Comment (yadudoc) [reply parsl/executors/flux/executor.py:154]: We should probably chat about this. We could essentially have the executor define a flux-specific `launch_cmd` that is then handed off to a parsl provider that will do the instantiation of Flux within a batch job.

File: parsl/executors/flux/executor.py
Comment (yadudoc) [reply parsl/executors/flux/executor.py:205]: You could use an atomic counter, and it's probably nicer than UUIDs.

File: parsl/executors/flux/executor.py
Comment (yadudoc) [reply parsl/executors/flux/executor.py:228]: I feel like I care about this because I've always had it these separate, but I do not know whether a user would care. Does flux default to merging the streams?

File: parsl/executors/flux/executor.py
Comment (jameshcorbett) [reply parsl/executors/flux/executor.py:154]: Right, I was thinking something like that would be a good idea. The actual command will depend on the resource manager, and possibly the platform, though.

For instance, on LLNL Slurm machines we have to add the ``--mpibind=off`` option when we launch Flux in order to get decent performance.

The goal is generally to launch a single instance of Flux on each node in a batch job.

The executor can also connect to a remote Flux instance, but I think LocalProvider would cover that use case.

File: parsl/executors/flux/execute_parsl_task.py
Comment (hategan) [reply parsl/executors/flux/execute_parsl_task.py:31]: I think this behavior is not consistent with the general contract of `Future`. The idea being that `Future.result()` should behave as if the computation wrapped by the future was invoked synchronously. That is, if said computation  throws an exception instead of returning a value, so should `Future.result()`.

Comment (jameshcorbett) [reply parsl/executors/flux/execute_parsl_task.py:31]: Right, I think we're all agreed on what's supposed to happen.

File: parsl/executors/flux/executor.py
Comment (hategan) [reply parsl/executors/flux/executor.py:205]: If you're grepping through a log file, there's no way a UUID is anything else but what it's supposed to be. If, on the other hand, you're grepping for `1`, you might get a few more extra hits.

File: parsl/executors/flux/executor.py
Comment (jameshcorbett) [reply parsl/executors/flux/executor.py:228]: No, Flux defaults to storing it in its KVS. I'm happy to make them separate, although I'm used to have them united on the terminal, which is why I united them here. I think we should do whatever Parsl does elsewhere, if it does anything similar.

File: parsl/executors/flux/executor.py
Comment (jameshcorbett) [reply parsl/executors/flux/executor.py:101]: Flux doesn't currently have any way to request additional nodes or resources, I believe, although I know that is an area of research for the Flux team. I think you would have to request more nodes from the local resource manager and then launch Flux in the new allocation.

I mentioned in another comment that the executor can connect to remote Flux instances, so it might not be too hard to have a executor which communicates with multiple allocations (and hence multiple Flux instances). But it would take some work to get the load-balancing right across allocations, unless Parsl already has code to handle that.

I think that functionality, though, like the functionality of having a hierarchy of Flux instances controlled by the executor, could be added later.

File: parsl/executors/high_throughput/process_worker_pool.py
Comment (jameshcorbett) [inline parsl/executors/high_throughput/process_worker_pool.py:467]: The exception was actually clobbered because ``logger`` wasn't defined. I thought the log message was not super useful anyway so I just removed it and the whole try/except block.

File: parsl/executors/flux/execute_parsl_task.py
Comment (jameshcorbett) [reply parsl/executors/flux/execute_parsl_task.py]: Added a commit that should resolve this problem.

File: parsl/executors/flux/executor.py
Comment (benclifford) [reply parsl/executors/flux/executor.py:106]: I've done some work to pull out HighThroughputExecutors block provisioning code into a superclass so it can also be used by the WorkQueueExecutor. The abstraction there is roughly that that kind of executor wants to execute task by having blocks of workers allocated by a batch system, but the nature of those workers is executor specific (so in the HightThroughputExecutor case, it is our process_worker_pool.py script; in the case of workqueue, it is workqueue's own `workqueue_worker` command. That abstraction also deals with deciding (in a simple way) when more blocks should be requested.

This abstraction doesn't prevent a user from saying "don't allocate any blocks at all - I will do this myself", which makes sense if a workqueue user wants to use some other provisioning tools provided by workqueue.

It sounds like potentially that abstraction fits your flux implementation too - if so, I can push on getting this into `master` a bit harder.

Comment (jameshcorbett) [reply parsl/executors/flux/executor.py:106]: That sounds great, I think that would fit well with this executor. However, what about load balancing? When a new app is submitted to the executor, how would the executor decide what blocks to forward the app onto? Is there any kind of abstraction for that or would that be up to me to implement? That could be pretty hard if each app has a variable amount of resources (cores and gpus).

Comment (benclifford) [reply parsl/executors/flux/executor.py:106]: > When a new app is submitted to the executor, how would the executor decide what blocks to forward the app onto?

In the HighThroughputExecutor/WorkQueue model, that is one of the primary purposes of the executor implementation - they have some component (different in each) that the workers all connect with.

Comment (jameshcorbett) [reply parsl/executors/flux/executor.py:106]: And the workers report to the executor how much they are being utilized, so the executor knows which worker to pass new apps to?

Comment (benclifford) [reply parsl/executors/flux/executor.py:106]: there's usually some kind of communication like that but it's executor-specific exactly what is going on there.

Comment (jameshcorbett) [reply parsl/executors/flux/executor.py:106]: Right, that makes sense. I'm just not sure there is great Flux functionality for figuring out whether a Flux instance is being underutilized or not. Off the top of my head the only thing I can think of is going by the number/percentage of PENDING jobs. I will talk to the Flux team about this.

File: parsl/executors/flux/executor.py
Comment (jameshcorbett) [reply parsl/executors/flux/executor.py:241]: We agreed in a conversation to separate them.

File: parsl/executors/flux/executor.py
Comment (yadudoc) [reply parsl/executors/flux/executor.py:236]: Let's punt on this until a user complains that they want to set cwd per app.

File: parsl/executors/flux/executor.py
Comment (jameshcorbett) [reply parsl/executors/flux/executor.py:108]: For reference: https://github.com/flux-framework/flux-core/discussions/3656

File: parsl/executors/flux/executor.py
Comment (jameshcorbett) [reply parsl/executors/flux/executor.py]: Based on our talk on Tuesday, I updated the implementation so that it properly supports Parsl's execution model and doesn't assume that it is running inside a Flux instance on the target resources. Depending on the setup, users may want to provider overrides to a Launcher instance, or to mess with the executor's ``launch_cmd``. I don't know if there will be any more user-friendly way of doing it, because it is so dependent on the specific platform (LLNL Slurm vs LANL Slurm, for instance).

**Comment by yadudoc** [comment]
@jameshcorbett Could you check the flake8 errors from CI here: https://travis-ci.com/github/Parsl/parsl/jobs/510817580#L544

**Comment by jameshcorbett** [comment]
I believe I have fixed all of the issues but I cannot quite get the CI to pass. Each time one of the builds seems to error out.

**Comment by benclifford** [comment]
> I cannot quite get the CI to pass. 

the last run seems to have passed!

The CI has a few tests that fail often in circumstances I've been unable to diagnose.

**Comment by yadudoc** [comment]
@jameshcorbett I just tested this out on quartz, and it looks like the tests are working! :tada: At this point I'd recommend merging this PR, and following it up with a separate documentation PR that covers when, where and how to use the `FluxExecutor`.

**Review by yadudoc — APPROVED** [review]
:+1:

File: parsl/tests/test_flux.py
Comment (benclifford) [inline parsl/tests/test_flux.py:20]: I'm not super fond of this style of magic autodetection of if a test should happen. Elsewhere we use explicit pytest `-k` arguments to disable tests we expect to not pass (see eg. `-k not cleannet and not issue363` in the Makefile). Every time I've encountered this style of testing, I've encountered tests incorrectly passing because this code will incorrectly not test flux when I've installed it incorrectly.

Comment (jameshcorbett) [reply parsl/tests/test_flux.py:20]: That makes sense to me. There are three separate sets of tests in the module:

1. Tests that can run any time
2. Tests that require an installation of flux to be in `PATH`
3. Tests that must be run inside of a Flux instance, e.g. `flux start python -m pytest test_module.py`

3 implies 2. Should I split 1 and (2, 3) into separate modules, and then add a target "flux_test" for explicitly checking tests of type 2 and 3?

Also I'm a little confused about `-k`/`-m` because it looks like (according to pytest documentation) `-k` is for matching tests by name and `-m` is for matching tests by marker, and yet `-k` seems to be used in Parsl for matching tests by marker... am I missing something?

Comment (benclifford) [reply parsl/tests/test_flux.py:20]: huh, interesting. I hadn't noticed that -k vs -m thing.

I think you should get this PR merged (by clicking "update from master" and then waiting for a parsl committer to merge it, without putting in any further changes), and then fiddle with tests separately.

I don't think the layout of the test modules matters too much - just ensure the ones that require flux behaviour have a proper tag attached to them to make them disabled in the CI where necessary. Adding a flux test makefile target would be ok too but not mandatory.

**Pull request merged**
